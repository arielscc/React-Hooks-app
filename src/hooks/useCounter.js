import { useState } from 'react';

export const useCounter = (initialState = 10) => {
  const [counter, setCounter] = useState(initialState);

  const increment = (e, value = 1) => {
    return setCounter(counter + value);
  };
  const decrement = (e, value = 1) => {
    return setCounter(counter - value);
  };

  const reset = () => {
    return setCounter(initialState);
  };

  return {
    counter,
    increment,
    decrement,
    reset,
  };
};
