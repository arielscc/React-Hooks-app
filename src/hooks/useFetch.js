import { useState, useEffect, useRef } from 'react';

export const useFetch = (url) => {
  const initialState = {
    data: null,
    loading: true,
    error: null,
  };

  const isMounted = useRef(true);

  const [state, setState] = useState(initialState);

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    setState(initialState);
    fetchData();
  }, [url]);

  const fetchData = () => {
    fetch(url)
      .then((resp) => resp.json())
      .then((data) => {
        setTimeout(() => {
          if (isMounted.current) {
            setState({
              data,
              loading: false,
              error: null,
            });
          }
        }, 500);
      })
      .catch(() => {
        setState({
          data: null,
          loading: false,
          error: 'No se pudo cargar la info',
        });
      });
  };

  return state;
};
