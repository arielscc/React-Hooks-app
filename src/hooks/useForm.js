import { useState } from 'react';

export const useForm = (initialState = {}) => {
  const [values, setValues] = useState(initialState);
  console.log('useform');
  const handleInputChange = ({ target }) => {
    setValues({
      ...values,
      [target.name]: target.value,
    });
  };

  const resetValues = () => setValues(initialState);

  return [values, handleInputChange, resetValues];
};
