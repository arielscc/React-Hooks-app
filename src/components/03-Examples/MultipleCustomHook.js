import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import { ThreeDots } from 'svg-loaders-react';
import { useCounter } from '../../hooks/useCounter';

export const MultipleCustomHook = (_) => {
  const { counter, increment } = useCounter();

  const { loading, data } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );

  const { author, quote } = !!data && data[0];

  return (
    <div className="w-4/5 m-auto text-gray-700">
      <h1 className="text-center text-4xl font-bold ">Braking Bad Quotes</h1>
      <hr />
      <div className="w-10/12 md:px-10 p-3  mx-auto mt-5 rounded-lg md:h-24 text-lg text-white italic flex items-center justify-center md:font-semibold bg-gradient-to-r from-teal-400 to-indigo-500 ">
        {loading ? (
          <ThreeDots fill="#5A67D8" className="h-4" />
        ) : (
          <p>{quote}</p>
        )}
      </div>
      <blockquote className="flex justify-end flex-col text-right w-11/12">
        <footer className="italic font-thin text-gray-600">{author}</footer>
      </blockquote>
      <button
        onClick={increment}
        className="bg-purple-600 hover:bg-purple-700 focus:outline-none focus:ring-2 text-white px-4 py-2 rounded-md"
      >
        siguiente
      </button>
    </div>
  );
};
