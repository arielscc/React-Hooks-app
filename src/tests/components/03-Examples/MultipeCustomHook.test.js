import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHook } from '../../../components/03-Examples/MultipleCustomHook';
import { useCounter } from '../../../hooks/useCounter';
import { useFetch } from '../../../hooks/useFetch';

jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('Pruebas en <MultipleCustomHook/>', () => {
  useCounter.mockReturnValue({
    counter: 1,
    increment: () => {},
  });

  test('Debe de mostrarse correctamente, Snapshot', () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });

    const wrapper = shallow(<MultipleCustomHook />);
    expect(wrapper).toMatchSnapshot();
  });

  test('debe de retornar un quote', () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: 'Ariel',
          quote: 'Hola a todos',
        },
      ],
      loading: false,
    });

    const wrapper = shallow(<MultipleCustomHook />);
    expect(wrapper.find('ThreeDots').exists()).toBe(false);
    expect(wrapper.find('p').text().trim()).toBe('Hola a todos');
    expect(wrapper.find('footer').text().trim()).toBe('Ariel');
  });

  test('Debe de mostrar el counter en 1', () => {});
});
