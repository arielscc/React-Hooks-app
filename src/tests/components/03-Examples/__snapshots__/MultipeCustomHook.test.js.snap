// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHook/> Debe de mostrarse correctamente, Snapshot 1`] = `
<div
  className="w-4/5 m-auto text-gray-700"
>
  <h1
    className="text-center text-4xl font-bold "
  >
    Braking Bad Quotes
  </h1>
  <hr />
  <div
    className="w-10/12 md:px-10 p-3  mx-auto mt-5 rounded-lg md:h-24 text-lg text-white italic flex items-center justify-center md:font-semibold bg-gradient-to-r from-teal-400 to-indigo-500 "
  >
    <A
      className="h-4"
      fill="#5A67D8"
    />
  </div>
  <blockquote
    className="flex justify-end flex-col text-right w-11/12"
  >
    <footer
      className="italic font-thin text-gray-600"
    />
  </blockquote>
  <button
    className="bg-purple-600 hover:bg-purple-700 focus:outline-none focus:ring-2 text-white px-4 py-2 rounded-md"
    onClick={[Function]}
  >
    siguiente
  </button>
</div>
`;
